### PR TITLE
Make make_labels() more consistent

### LINKED ISSUE #2975: ggplot adds backticks to labels (regression?)
ggplot adds backticks to labels (see x-axis label). The behavior was different in previous versions (e.g., 2.2.1).

``` r
library(ggplot2)
d <- data.frame(rnorm(100))
colnames(d) <- 'a b'
ggplot(d, aes(`a b`)) + geom_histogram()
#> `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.
```

Created on 2018-11-01 by the [reprex package](http://reprex.tidyverse.org) (v0.2.0).

### PR DESCRIPTION
**Pull request by yutannihilation**
Fix #2975 

### Problem

Currently, ggplot2 makes labels in three different ways:

1. `make_labels()` uses `quo_text()`. This handles the cases where `aes()` is passed to `ggplot()` or `geom_*()`.
2.  [`ggplot_add.uneval()`](https://github.com/tidyverse/ggplot2/blob/1c09bae2aa24320bcb4891c664cccc16efc86a8a/R/plot-construction.r#L126) uses `quo_name()`. This handles the case where `aes()` is `+`ed to the plot.
3. [`qplot()`](https://github.com/tidyverse/ggplot2/blob/1c09bae2aa24320bcb4891c664cccc16efc86a8a/R/quick-plot.r#L90) uses `quo_name()`, (but the implementation is yet different from `ggplot_add.uneval()`). This handles `qplot()`.

Consequently, we see some inconsistencies among them. For example,

* `make_labels()` adds backticks to invalid symbols (e.g. `` `a b` ``), while others don't.
* If the mapping is `NULL`, `make_labels()` uses the aesthetic name, `ggplot_add.uneval()` removes the label, and `qplot()` uses `"NULL"`.

### Scope of this PR

The backticks can be easily removed if we stop using `quo_text()` for symbols. So, at least we can and should fix `make_labels()`.

Next, ideally, we should use `make_label()` for making labels everywhere. But, in order to keep this PR small, this PR does only these:

* Make `make_labels()` to follow the disirable behaviours (below).
* Use `make_labels()` in `ggplot_add.uneval()`.

and I leave these issues:

* `qplot()` uses different implementation
* Facetting also uses different implementation, which uses `deparse()` (https://github.com/tidyverse/ggplot2/issues/2975#issuecomment-436285746)

### Desirable Behaviours

* For quosures of a symbol, use symbols as is whether or not the symbol is valid by the R's parser (i.e. do not add backticks). Use `as_string(quo_get_expr())` for this.
* For other types of quosures, deparse them as R's expressions (i.e. add backticks if necessary). If the expression is long, the label should be abbreviated with a tailing `...`. Use `quo_text()` and `gsub()` for this.
* For literals, use the aesthetic name instead.
* For `NULL`, use the aesthetic name instead.

### CODE REVIEW

**Comment by lionel-** [comment]
I think it makes sense, but the abbreviation of the calls may cause trouble in revdeps and production plots. Should you maybe run the revdeps to get a sense of how this impacts backward compatibility?

Or you could implement `aes_deparse_name()` which would use `as_string()` for symbols and `quo_text()` for everything else, that might be safer.

**Comment by yutannihilation** [comment]
Thanks, let's care about backward compatibility.

> use `as_string()` for symbols

In this case, it's not symbols; I want to convert quosures of a symbol to strings (my description above is a bit confusing, sorry). What is the supposed way to use `as_string()` for quosures? `as_string(quo_get_expr())`? Or, can I just use `quo_name()` after checking with `quo_is_symbol()`?

**Comment by clauswilke** [comment]
This seems to work.

``` r
make_labels2 <- function(mapping) {
  default_label <- function(aesthetic, mapping) {
    # e.g., geom_smooth(aes(colour = "loess"))
    if (is.atomic(mapping)) {
      aesthetic
    } else if (rlang::quo_is_symbol(mapping)) {
      rlang::as_string(ggplot2:::strip_dots(rlang::get_expr(mapping)))
    } else {
      x <- rlang::quo_text(ggplot2:::strip_dots(mapping))
      if (length(x) > 1) {
        x <- paste0(x[[1]], "...")
      }
      x
    }
  }
  Map(default_label, names(mapping), mapping)
}

m <- ggplot2::aes(`a b`)
ggplot2:::make_labels(m)
#> $x
#> [1] "`a b`"
make_labels2(m)
#> $x
#> [1] "a b"

m <- ggplot2::aes(2*x*exp(`coef 1`*x^2))
ggplot2:::make_labels(m)
#> $x
#> [1] "2 * x * exp(`coef 1` * x^2)"
make_labels2(m)
#> $x
#> [1] "2 * x * exp(`coef 1` * x^2)"
```

<sup>Created on 2018-11-05 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by lionel-** [comment]
Instead of `get_expr()` I would use `quo_get_expr()` because it fails early if the input type is not right.

**Comment by yutannihilation** [comment]
Thanks! I would use the `quo_get_expr()` version of @clauswilke's implementation.

BTW, is it OK to remove this `if` branch? I don't understand what case the current code considers since `rlang::quo_text()` always returns a single string (right?).

``` r
      if (length(x) > 1) {
        x <- paste0(x[[1]], "...")
      }
```

**Comment by clauswilke** [comment]
I edited my previous comment to include an example implementation of what I meant.

**Comment by lionel-** [comment]
In rlang I also see:

```r
name <- quo_text(expr)
name <- gsub("\n.*$", "...", name)
```

**Comment by yutannihilation** [comment]
Ah, got it! Thanks. I didn't know the difference between `quo_text()` and `deparse()`...

Note that, to be precise about the backward compatibility, it seems that `...` had never appeared in the release versions of ggplot2. Version 2.2.1 uses deparsed string as is (https://github.com/tidyverse/ggplot2/blob/v2.2.1/R/labels.r#L86):

``` r
# Convert aesthetic mapping into text labels
make_labels <- function(mapping) {
  remove_dots <- function(x) {
    gsub(match_calculated_aes, "\\1", x)
  }

  default_label <- function(aesthetic, mapping) {
    # e.g., geom_smooth(aes(colour = "loess"))
    if (is.character(mapping)) {
      aesthetic
    } else {
      remove_dots(deparse(mapping))
    }
  }
  Map(default_label, names(mapping), mapping)
}
```

I don't follow the codes detailedly, but it seems only the first element is used for labeling.

``` r
library(ggplot2)

packageVersion("ggplot2")
#> [1] '2.2.1'

d <- data.frame(a = 1)

ggplot(d, aes(a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a + a)) +
  geom_bar()
```



<sup>Created on 2018-11-06 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

But, using `gsub()` does the right thing to me. So, I think I'm going to implement as such if I find some reason not to :)

**Comment by yutannihilation** [comment]
For `ggplot_add.uneval()`, I chose to follow https://github.com/tidyverse/ggplot2/issues/2975#issuecomment-436079361 and use `make_labels()` as well.

For `qplot()`, I'm failing to see what is the expected way to make labels, so I want to leave them as is for now...

``` r
library(ggplot2)

patchwork::wrap_plots(
  qplot(x = 1, geom = "vline"),
  qplot(x = 1, y = NULL, geom = "vline"),
  ggplot(mapping = aes(x = 1)) + geom_vline(),
  ggplot(mapping = aes(x = 1, y = NULL)) + geom_vline()
)
```



<sup>Created on 2018-11-06 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by yutannihilation** [comment]
Hmm, it seems the handling of `NULL` diverged after https://github.com/tidyverse/ggplot2/pull/2617, but maybe `make_labels(aes(x = NULL))` should return `NULL`?

**Comment by lionel-** [comment]
Perhaps this patch should make minimal changes and we'll solve the problem more fully as part of https://github.com/r-lib/rlang/issues/636

**Comment by hadley** [comment]
At a minimum, we should write down the labelling principles that ggplot2 wants, so we can make sure to enforce them in rlang. I think the reason that `quo_name()`, `quo_text()` etc keep changing is that I failed to properly analyse and document the use cases.

**Comment by clauswilke** [comment]
I think `NULL` was handled correctly before. The principle seems to be that we use the aesthetic name when a specific value is provided. This makes sense, because we want the x axis to be called x, not 15. Similarly, we also don‘t want the x axis to be called NULL.

**Comment by yutannihilation** [comment]
> I think `NULL` was handled correctly before.

Now I gradually feel you are right, but let me confirm what you mean by "before". Currently, ggplot2 handles `NULL` in three different ways:

1. `make_labels()` uses the aesthetic name instead for the label
2. `ggplot_add.uneval()` displays no label (introduced by #2617)
3. `qplot()` uses `"NULL"` for the label

Do you mean 1. is the supposed behaviour? (I think we can agree 3. is not the answer at least)

``` r
library(ggplot2)

patchwork::wrap_plots(
  ggplot(mapping = aes(x = 1, y = NULL)) + geom_vline() + ggtitle("make_labels()"),
  ggplot() + aes(x = 1, y = NULL) + geom_vline() + ggtitle("ggplot_add.uneval()"),
  qplot(x = 1, y = NULL, geom = "vline") + ggtitle("qplot()")
)
```



<sup>Created on 2018-11-07 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by hadley** [comment]
I'd consider 1 to be the desired behaviour. 3 is bad, but it's not necessary to fix it (unless it's easy)

**Comment by clauswilke** [comment]
Yes, I think option 1 is the most likely to make sense in an ambiguous situation.

``` r
library(ggplot2)
ggplot(mapping = aes(x = 1, y = NULL)) + geom_point(data = data.frame(x = 1, y = 1), mapping = aes(x, y))
```



``` r
ggplot() + aes(x = 1, y = NULL) + geom_point(data = data.frame(x = 1, y = 1), mapping = aes(x, y))
```



<sup>Created on 2018-11-06 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by yutannihilation** [comment]
Thanks! OK, I'll modify the codes to follow option 1.

**Comment by yutannihilation** [comment]
I updated the description of this PR. Does this correctly reflect our discussion so far?

**Comment by clauswilke** [comment]
Yes, I think so.

File: tests/testthat/test-aes-calculated.r [hunk 1]
```diff
@@ -24,7 +24,17 @@ test_that("strip_dots remove dots around calculated aesthetics", {
...
+test_that("make_labels() deprases mappings properly", {
+  # calculation stripped from labels
+  expect_identical(make_labels(aes(x = ..y..)), list(x = "y"))
+  expect_identical(make_labels(aes(x = stat(y))), list(x = "y"))
+
+  # symbol is always deparsed without backticks
+  expect_identical(make_labels(aes(x = `a b`)), list(x = "a b"))
+  # long expression is abbreviated with ...
+  expect_identical(make_labels(aes(x = 2 * x * exp(`coef 1` * x^2) * 2 * x * exp(`coef 1` * x^2) * 2 * x)),
+                   list(x = "2 * x * exp(`coef 1` * x^2) * 2 * x * exp(`coef 1` * x^2) * 2 * ..."))
```
[the remark this answers is not in the document]
Comment (yutannihilation) [reply tests/testthat/test-aes-calculated.r]: Sure.

**Comment by yutannihilation** [comment]
@clauswilke @dpseidel May I merge this now? If you are planning a quick release with the fix for the transformation of secondary axis, I think I should wait. Though this PR is only about fixing a minor bug, worries about breaking revdeps should be minimized...

**Comment by clauswilke** [comment]
I think this should be merged now. The change in rlang had much bigger implications for revdeps, and this PR mostly undoes that change.

**Comment by yutannihilation** [comment]
> The change in rlang had much bigger implications for revdeps, and this PR mostly undoes that change.

Thanks, makes sense!

**Pull request merged**